File: entrylens-api/app/routes/labs.py
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from fastapi.responses import FileResponse

from app.auth import verify_api_key
from app.services.labs import (
    get_lab_state,
    resolve_lab_file,
    run_playground_command,
    save_enrollment_images,
    save_probe_image,
)

router = APIRouter(tags=["labs"])


@router.get("", dependencies=[Depends(verify_api_key)])
async def lab_state(target: str = "mediapipe") -> dict:
    return get_lab_state(target=target)


@router.get("/files", dependencies=[Depends(verify_api_key)])
async def lab_file(target: str, file_path: str) -> FileResponse:
    return FileResponse(resolve_lab_file(target, file_path))


@router.post("/enroll-images", dependencies=[Depends(verify_api_key)])
async def upload_enroll_images(
    person_name: Annotated[str, Form()],
    files: Annotated[list[UploadFile], File()],
    target: Annotated[str, Form()] = "local-recognition",
) -> dict:
    return await save_enrollment_images(target, person_name, files)


@router.post("/probe-image", dependencies=[Depends(verify_api_key)])
async def upload_probe_image(
    file: Annotated[UploadFile, File()],
    probe_name: Annotated[str | None, Form()] = None,
    target: Annotated[str, Form()] = "mediapipe",
) -> dict:
    return await save_probe_image(target, file, probe_name)


@router.post("/commands/{command}", dependencies=[Depends(verify_api_key)])
async def execute_playground_command(
    command: str,
    target: str = "mediapipe",
    file_path: str | None = None,
    person_name: str | None = None,
) -> dict:
    try:
        return await run_playground_command(target, command, file_path=file_path, person_name=person_name)
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Vision lab command failed before execution: {exc}",
        ) from exc

File: entrylens-frontend/src/api/labs.ts
import { apiFetch, apiFetchBlob } from "./client";
import type { LabStatePayload } from "../types";

export function getLabsState(target = "mediapipe"): Promise<LabStatePayload> {
  return apiFetch<LabStatePayload>(`/api/v1/labs?target=${encodeURIComponent(target)}`);
}

export function uploadEnrollImages(personName: string, files: File[], target = "local-recognition"): Promise<{ saved_files: string[]; person: string }> {
  const formData = new FormData();
  formData.append("person_name", personName);
  formData.append("target", target);
  for (const file of files) {
    formData.append("files", file);
  }

  return apiFetch<{ saved_files: string[]; person: string }>("/api/v1/labs/enroll-images", {
    method: "POST",
    body: formData,
  });
}

export function uploadProbeImage(file: File, probeName = "", target = "mediapipe"): Promise<{ probe_file: string }> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("target", target);
  if (probeName) {
    formData.append("probe_name", probeName);
  }

  return apiFetch<{ probe_file: string }>("/api/v1/labs/probe-image", {
    method: "POST",
    body: formData,
  });
}

export function runLabsCommand(
  command: string,
  options: { target?: string; filePath?: string; personName?: string } = {},
): Promise<Record<string, unknown>> {
  const params = new URLSearchParams();
  params.set("target", options.target || "mediapipe");
  if (options.filePath) {
    params.set("file_path", options.filePath);
  }
  if (options.personName) {
    params.set("person_name", options.personName);
  }

  const suffix = params.toString() ? `?${params.toString()}` : "";
  return apiFetch<Record<string, unknown>>(`/api/v1/labs/commands/${command}${suffix}`, {
    method: "POST",
  });
}

export function fetchLabsFileBlob(target: string, filePath: string): Promise<Blob> {
  const params = new URLSearchParams();
  params.set("target", target);
  params.set("file_path", filePath);
  return apiFetchBlob(`/api/v1/labs/files?${params.toString()}`);
}
